File: interface/plant/py/hardware/plant.py
import sys
# sys.path.append(r"C:\Quanser\0_libraries\python")
from pal.products.aero2 import Aero2
from pal.utilities.scope import MultiScope

import tcp_protocol_server as tcs

# init tcp host and port
HOST = '0.0.0.0'
PORT = 9999

# get other tools
from threading import Thread
import signal
import time
import math
import numpy as np

# Thread hanlder initialization
global KILL_THREAD
KILL_THREAD = False
def sig_handler(*args):
    global KILL_THREAD
    KILL_THREAD = True
signal.signal(signal.SIGINT, sig_handler)

# simulation time and plotting set
simulationTime = 30 # will run for 30 seconds
color = np.array([0, 1, 0], dtype=np.float64)

Scope = MultiScope(rows=3, cols=2, title='Aero2', fps=30)

Scope.addAxis(row=0, col=0, timeWindow=12, yLabel='Pitch (deg)')
Scope.axes[0].attachSignal(name='theta_p')
Scope.axes[0].attachSignal(name='ref_p')

Scope.addAxis(row=0, col=1, timeWindow=12, yLabel='Yaw (deg)')
Scope.axes[1].attachSignal(name='theta_y')
Scope.axes[1].attachSignal(name='ref_y')

Scope.addAxis(row=1, col=0, timeWindow=12, yLabel='Pitch Rate (deg/s)')
Scope.axes[2].attachSignal(name='dp')

Scope.addAxis(row=1, col=1, timeWindow=12, yLabel='Yaw Rate (deg/s)')
Scope.axes[3].attachSignal(name='dy')

Scope.addAxis(row=2, col=0, timeWindow=12, xLabel='Time (s)', yLabel='V_main (V)')
Scope.axes[4].attachSignal(name='V_main')

Scope.addAxis(row=2, col=1, timeWindow=12, xLabel='Time (s)', yLabel='V_tail (V)')
Scope.axes[5].attachSignal(name='V_tail')

# control-system scenario
def control_loop():
    # interface setting #
    # ------------------------------------------------ #
    # using hardware
    id = 0
    
    # if you want to use Ouanser Interactive Labs, you will change to 0
    hardware = 0
    
    readMode = 0

    # frequency of system holder and sampler
    frequency = 50 # hz

    # for scope sampling rate
    countMax = frequency / 50
    count = 0

    # class initialization
    AeroClass = Aero2

    # describe #
    # ------------------------------------------------ #
    # instance of hardware model 
    with AeroClass(id, hardware=hardware, readMode=readMode, frequency=frequency) as myAero:
        # instance of tcp layer
        with tcs.tcp_server(HOST, PORT) as tcsp:
            startTime = 0
            timeStamp = 0
            def elapsed_time():
                return time.time() - startTime
            startTime = time.time()

            while timeStamp < simulationTime and not KILL_THREAD:
                # running signal send for controller
                tcsp.send("run")

                # read sensor information
                myAero.read_analog_encoder_other_channels()

                # calc output
                theta_pitch = myAero.pitchAngle
                theta_yaw = myAero.yawAngle
                rate_pitch = myAero.pitchRate
                rate_yaw = myAero.yawRate

                # send plant output
                tcsp.send(theta_pitch)
                tcsp.send(theta_yaw)
                tcsp.send(rate_pitch)
                tcsp.send(rate_yaw)

                # get control input
                _, u_motr_1 = tcsp.recv()
                _, u_motr_2 = tcsp.recv()

                # get ref value
                _, ref_pitch = tcsp.recv()
                _, ref_yaw = tcsp.recv()

                # running range set
                if abs(np.rad2deg(theta_pitch)) > 40:
                    voltage_motr_1, voltage_motr_2 = 0.0, 0.0
                else:
                    voltage_motr_1, voltage_motr_2 = u_motr_1, u_motr_2

                # write commands
                # voltage = np.clip(voltage, -15, 15)
                myAero.write_voltage(voltage_motr_1, voltage_motr_2)

                # plot to scopes
                count += 1
                if count >= countMax:
                    Scope.axes[0].sample(timeStamp, [np.rad2deg(theta_pitch), np.rad2deg(ref_pitch)])
                    Scope.axes[1].sample(timeStamp, [np.rad2deg(theta_yaw), np.rad2deg(ref_yaw)])
                    Scope.axes[2].sample(timeStamp, [np.rad2deg(rate_pitch)])
                    Scope.axes[3].sample(timeStamp, [np.rad2deg(rate_yaw)])
                    Scope.axes[4].sample(timeStamp, [u_motr_1])
                    Scope.axes[5].sample(timeStamp, [u_motr_2])
                    count = 0

                timeStamp = elapsed_time()

            tcsp.send("end")

def main():
    thread = Thread(target=control_loop)
    thread.start()

    while thread.is_alive() and (not KILL_THREAD):

        # This must be called regularly or the scope windows will freeze
        # Must be called in the main thread.
        MultiScope.refreshAll()
        time.sleep(0.01)

    input('Press the enter key to exit.')

if __name__ == "__main__":
    main()

File: interface/plant/cpp/hardware/plant.cpp
#define _USE_MATH_DEFINES
#include <iostream>
#include <cstdint>
#include <string>
#include <cmath>
#include <chrono>
#include "tcp_protocol_server_windows.h"
#include "hil.h"
#include "quanser_timer.h"
using namespace std;

const string host = "0.0.0.0";
const int port = 9999;

int main()
{
    t_card board;
    t_error result;
    // if you want to use hardware change to 1 else 0 is virtual(QLab)
    bool hardware = 0;

    if (hardware)
    {
        result = hil_open("quanser_aero2_usb", "0", &board);
        if (result < 0)
        {
            cout << "failure to connect hardware" << endl;
            return -1;
        }
    }
    else
    {
        result = hil_open("quanser_aero2_usb", "0@tcpip://localhost:18950", &board);
        if (result < 0)
        {
            cout << "failure to connect QLab" << endl;
            return -1;
        }
    }

    // simulation_time is total run time, sample_time is sample_time.
    int simulation_time = 30;
    double sample_time = 0.1;
    t_timeout interval;
    t_timeout timeout;
    timeout_get_timeout(&interval, sample_time);
    timeout_get_current_time(&timeout);

    double state[4] = { 0.0, 0.0, 0.0, 0.0 };
    double voltage[2] = { 0.0, 0.0 };
    double ref[2] = { 0.0, 0.0 };
    int32_t encoder_counts[4];
    uint32_t encoder_channels[4] = { 0, 1, 2, 3 };
    double other_counts[10];
    uint32_t other_channles[10] = { 3000, 3001, 3002, 4000, 4001, 4002, 14000, 14001, 14002, 14003 };
    uint32_t analog_channels[2] = { 0, 1 };
    uint32_t digital_channels[2] = { 0, 1 };
    t_boolean digital_values[2] = { 1, 1 };
    hil_write_digital(board, digital_channels, 2, digital_values);

    // calculation angle
    double theta_pitch = 0.0;
    double theta_yaw = 0.0;
    double rate_pitch = 0.0;
    double rate_yaw = 0.0;


    // TCP/IP ready
    tcp_server tcsp = tcp_server(host, port);

    // Just check loop/total time
    auto stc = chrono::high_resolution_clock::now();
    auto edc = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::nanoseconds>(edc - stc);
    double run_time = duration.count() / 1000000;
    double stack_time = 0.0;


    // control loop
    for (int64_t i = 0; i < (int64_t)((double)simulation_time / sample_time); i++)
    {
        stc = chrono::high_resolution_clock::now();

        // time sleep
        timeout_add(&timeout, &timeout, &interval);
        qtimer_sleep(&timeout);

        // read sensor
        hil_read_encoder(board, encoder_channels, 4, encoder_counts);
        hil_read_other(board, other_channles, 10, other_counts);
        theta_pitch =  encoder_counts[2] * (2.0 * M_PI / 2880);
        theta_yaw = encoder_counts[3] * (2.0 * M_PI / 4096);
        rate_pitch = other_counts[8] * (2.0 * M_PI / 2880);
        rate_yaw = other_counts[9] * (2.0 * M_PI / 4096);

        // running signal send for controller
        tcsp.Send<string>("run");

        // send plant output
        tcsp.Send<double>(theta_pitch);
        tcsp.Send<double>(theta_yaw);
        tcsp.Send<double>(rate_pitch);
        tcsp.Send<double>(rate_yaw);

        // get control input
        voltage[0] = tcsp.Recv<double>();
        voltage[1] = tcsp.Recv<double>();

        // get ref value
        ref[0] = tcsp.Recv<double>();
        ref[1] = tcsp.Recv<double>();

        // running range set
        if (abs(theta_pitch * 180 / M_PI) > 40)
        {
            voltage[0] = 0.0;
            voltage[1] = 0.0;
        }


        // actuator write
        if (voltage[0] > 15) 
        {
            voltage[0] = 15;
        }
        else if (voltage[0] < -15)
        {
            voltage[0] = -15;
        }
        if (voltage[1] > 15)
        {
            voltage[1] = 15;
        }
        else if (voltage[1] < -15)
        {
            voltage[1] = -15;
        }

        hil_write_analog(board, analog_channels, 2, voltage);

        edc = chrono::high_resolution_clock::now();
        duration = chrono::duration_cast<chrono::nanoseconds>(edc - stc);
        run_time = duration.count() / 1000000;
        stack_time += run_time / 1000;
        cout << " --------------------------------------------------------------------------- " << endl;
        cout << "iter: " << i << " | loop time: " << run_time << "ms | total time: " << stack_time << "s" << endl;
        cout << "motor 1 volt: " << voltage[0] << " | motor 2 volt: " << voltage[1] << endl;
        cout << "pitch: " << theta_pitch * 180 / M_PI << " | yaw: " << theta_yaw * 180 / M_PI << " | pitch_ref: " << ref[0] * 180 / M_PI << " | yaw_ref: " << ref[1] * 180 / M_PI << endl;
    }

    tcsp.Send<string>("end");

    // logic terminate
    voltage[0] = 0.0;
    voltage[1] = 0.0;
    hil_write_analog(board, analog_channels, 2, voltage);
    digital_values[0] = 0;  digital_values[1] = 0;
    hil_write_digital(board, digital_channels, 2, digital_values);
    if (board != NULL)
    {
        hil_close(board);
    }

    return 0;
}
